### PR TITLE
[MRG] Perfusion

### PR DESCRIPTION
**Pull request by nadkarni-na**
Found ways to translate it reasonably well into python, or at least more easily than passing data structures to the original R code would have been.

Code is parallelized per-voxel and has a progress bar. It has been tested on several images and produces reasonable results, though have not yet compared to R results..

There are lots of potential issues around the non-linear optimisation (see docstrings and comments). The easiest to resolve is probably to add a jacobian, which is apparently possible for the function we use.

Not sure how to best deal with https://docs.python.org/2/library/multiprocessing.html#windows section "Safe importing of the main module". In python 2 it's considered a Windows-only issue but in python 3 it's considered good practice to protect regardless of OS. Not sure how to make this transparent to the end user.

Not sure what version of tqdm to have as minimum.

The struct.py modification is a mistake in my branch control, will deal with it later.

### CODE REVIEW

**Comment by salma1601** [comment]
Great !!! Ping me when it is ready for a review.

**Comment by nadkarni-na** [comment]
It's worth starting now

**Comment by nadkarni-na** [comment]
Test data is https://www.dropbox.com/s/j8zvpr674xdwrx3/MLRS4-PERF-3.zip?dl=0
Not all of it produces good results, but that is due to acquisition problems

File: sammba/modality_processors/perfusion_fair.py [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+import numpy as np
+from scipy.optimize import least_squares as ls
+from multiprocessing import Pool
+from functools import partial
+import tqdm
+
+
+def perf_fair_read_ptbl(perf_fair_fname):
+    """
+    Extract important information about the perfusion FAIR acquisition
```
Comment (salma1601) [inline sammba/modality_processors/perfusion_fair.py]: can you be more explicit about *important* information

File: sammba/modality_processors/perfusion_fair.py [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+        *.nii.gz.
+
+    Returns
+    -------
+    Dictionary of parameters useful to perfusion fitting functions.    
+    """
+    
+    # the *_ptbl.txt file will have the same name as the perfusion file itself,
+    # but with _ptbl.txt at the end rather than just .nii.gz. arguably this
+    # function should directly take the text file name as input, but it is
```
Comment (salma1601) [inline sammba/modality_processors/perfusion_fair.py]: I prefer keeping things simple. IMHO a function reading a ptbl.txt file should take this file as input. Can't see a reason not to do so

File: sammba/modality_processors/perfusion_fair.py [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+    # difficult to envisage a scenario where this would be needed. the
+    # *_ptbl.txt file has a specific format created by
+    # sammba.io_conversions.dcm_to_nii. if someone supplies their own file
+    # name, it would mean they changed the *_ptbl.txt name (why?) or they have
+    # their own formatted text file which this function would not be able to
+    # process
+    
+    bfx = perf_fair_fname.split('_')[-1].split('.')[0]
+    ptbl = pd.read_table(perf_fair_fname.replace(bfx + '.nii.gz', bfx + '_ptbl.txt'))
+    ti = ptbl.TI[ptbl.slice == 1][ptbl.FC == 'Selective Inversion'].tolist()
```
Comment (salma1601) [inline sammba/modality_processors/perfusion_fair.py]: may be at some point a check function could be coded, that makes sure fields `TI`, `slice` etc exist and throws a meaningful error instead. This will avoid weird errors if parsing an uncorrect file

File: sammba/modality_processors/perfusion_fair.py [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+    ----------
+    filename : str
+        Path to the perfusion *.nii.gz file converted by
+        sammba.io_conversions.dcm_to_nii, NOT to the *_ptbl.txt file itself.
+        The *_ptbl.txt file name is determined automatically fro; that of the
+        *.nii.gz.
+
+    Returns
+    -------
+    Dictionary of parameters useful to perfusion fitting functions.
```
Comment (salma1601) [inline sammba/modality_processors/perfusion_fair.py]: Could you please be more explicit about the dictionary keys ?

File: sammba/modality_processors/perfusion_fair.py [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+
+def perf_fair_read_ptbl(perf_fair_fname):
+    """
+    Extract important information about the perfusion FAIR acquisition
+    parameters from the *_ptbl.txt file created by
+    sammba.io_conversions.dcm_to_nii.
+
+    Parameters
+    ----------
+    filename : str
```
Comment (salma1601) [inline sammba/modality_processors/perfusion_fair.py]: here the parameter is called `perf_fair_fname` not `filename`

File: sammba/modality_processors/perfusion_fair.py [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+    fc = ptbl.FC[ptbl.slice == 1].tolist()
+    picker_sel = [n for n,x in enumerate(fc) if x == 'Selective Inversion']
+    picker_nonsel = [n for n,x in enumerate(fc) if x == 'Non-selective Inversion']
+    return {'TI':ti, 'long_TI':long_ti, 'FC':fc,
+            'picker_sel':picker_sel, 'picker_nonsel':picker_nonsel}
+
+
+#the perfusion fluid-attenuated inversion-recovery function
+#the jacobian should be callable but I do not know enough maths to create it
+def _fair_t1_func(x, ti, s0):
```
Comment (salma1601) [inline sammba/modality_processors/perfusion_fair.py]: could you please give `x` a more explicit name ?

File: sammba/modality_processors/perfusion_fair.py [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+        constant; an example value is 2800 at 11.7T (Duong ref).
+        
+    lambda_blood : int or float
+        The assumed blood tissue partition coefficient of water in ml per g.
+        Empirical constant; usually 0.9.
+        
+    ti : list of int or float
+        The inversion times. Must have the same length as s0.
+        
+    multiplier : int or float
```
Comment (salma1601) [inline sammba/modality_processors/perfusion_fair.py]: may be this parameter can be removed and tell instead that you are returning CBF in ml per g per ms. Users interested in another limit can do the conversion themselves.

File: sammba/modality_processors/perfusion_fair.py [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+    """
+    
+    # there are no standard errors as these are not in the current output of
+    # scipy.optimize.least_squares. the lmfit library can produce them, but
+    # during testing they were usually incalculable, and lmfit was twice as
+    # slow (or maybe I was not using it properly?) it is unlikely that the
+    # errors are of much use any way so will stay with
+    # scipy.optimize.least_squares.
+    
+    if outtype not in ['simple', 'complicated']: return 'unrecognized outtype'
```
Comment (salma1601) [inline sammba/modality_processors/perfusion_fair.py]:
```Python
if condition:
    raise ValueError("unrecognized outtype")
```

File: sammba/modality_processors/perfusion_fair.py [hunk 9]
```diff
@@ -0,0 +1,369 @@
...
+    # sammba.io_conversions.dcm_to_nii. if someone supplies their own file
+    # name, it would mean they changed the *_ptbl.txt name (why?) or they have
+    # their own formatted text file which this function would not be able to
+    # process
+    
+    bfx = perf_fair_fname.split('_')[-1].split('.')[0]
+    ptbl = pd.read_table(perf_fair_fname.replace(bfx + '.nii.gz', bfx + '_ptbl.txt'))
+    ti = ptbl.TI[ptbl.slice == 1][ptbl.FC == 'Selective Inversion'].tolist()
+    long_ti = ptbl.TI[ptbl.slice == 1].tolist()
+    fc = ptbl.FC[ptbl.slice == 1].tolist()
```
Comment (salma1601) [inline sammba/modality_processors/perfusion_fair.py]: are all these guys np.arrays? if so, may be it would be better keep them so, and use np.where in the next line to find the indices

File: sammba/modality_processors/perfusion_fair.py [hunk 10]
```diff
@@ -0,0 +1,369 @@
...
+    
+    picker_sel : list or numpy array of int
+        Vector indicating positions of selectively inverted signals in s0. The
+        lengths of picker_sel and picker_sel must sum to the length of s0.
+        
+    picker_nonsel : list or numpy array of int
+        Vector indicating positions of non-selectively inverted signals in s0.
+        The lengths of picker_sel and picker_sel must sum to the length of s0.
+        
+    outtype : {'simple', 'complicated'}, optional
```
Comment (salma1601) [inline sammba/modality_processors/perfusion_fair.py]: I always prefer having one unique output type if it is possible

File: sammba/modality_processors/perfusion_fair.py [hunk 11]
```diff
@@ -0,0 +1,369 @@
...
+    Returns
+    -------
+    List of all_s0.shape[0] lists, each of 8 floats (see perf_fair_fitter).
+    """
+    
+    # inspiration provided by:
+    # https://stackoverflow.com/q/5442910
+    # https://stackoverflow.com/a/45276885
+    
+    # if __name__ == '__main__': this might be dangerous
```
Comment (salma1601) [inline sammba/modality_processors/perfusion_fair.py]: I think this comment can be removed

File: sammba/modality_processors/perfusion_fair.py [hunk 12]
```diff
@@ -0,0 +1,369 @@
...
+    if 2 > 1: # to keep the indent in case above line is reinstated
```
Comment (salma1601) [inline sammba/modality_processors/perfusion_fair.py]: what do you mean ?

File: sammba/modality_processors/perfusion_fair.py [hunk 13]
```diff
@@ -0,0 +1,369 @@
...
+        constant; an example value is 2800 at 11.7T (Duong ref).
+        
+    lambda_blood : int or float
+        The assumed blood tissue partition coefficient of water in ml per g.
+        Empirical constant; usually 0.9.
+        
+    ti : list of int or float
+        The inversion times. Must have the same length as s0.
+        
+    multiplier : int or float
```
Comment (salma1601) [inline sammba/modality_processors/perfusion_fair.py]: same as previous comment

File: sammba/modality_processors/perfusion_fair.py [hunk 14]
```diff
@@ -0,0 +1,369 @@
...
+    all_s0 : 2D numpy array of int or float
+        The aquired signals. The first dimension is the number of s0 vectors,
+        the second dimension the length of all of them. So for a real image,
+        usually len(D1) >> len(D2).
+        
+    t1_blood : int or float
+        T1 of blood in ms at the acquisition field strength. Empirical
+        constant; an example value is 2800 at 11.7T (Duong ref).
+        
+    lambda_blood : int or float
```
Comment (salma1601) [inline sammba/modality_processors/perfusion_fair.py]: may be can be an optional parameter with 0.9 as default value

File: sammba/modality_processors/perfusion_fair.py [hunk 15]
```diff
@@ -0,0 +1,369 @@
...
+    
+    picker_sel : list or numpy array of int
+        Vector indicating positions of selectively inverted signals in s0. The
+        lengths of picker_sel and picker_sel must sum to the length of s0.
+        
+    picker_nonsel : list or numpy array of int
+        Vector indicating positions of non-selectively inverted signals in s0.
+        The lengths of picker_sel and picker_sel must sum to the length of s0.
+        
+    ncpu : int
```
Comment (salma1601) [inline sammba/modality_processors/perfusion_fair.py]: I think it should better be optional with default either to use all cpus or only 1

File: sammba/modality_processors/perfusion_fair.py [hunk 16]
```diff
@@ -0,0 +1,369 @@
...
```
Comment (salma1601) [inline sammba/modality_processors/perfusion_fair.py]: same as before

File: sammba/modality_processors/perfusion_fair.py [hunk 17]
```diff
@@ -0,0 +1,369 @@
...
+                       multiplier, t1_guess, picker_sel, picker_nonsel, ncpu):
+    """
+    Wrapper to execute perf_fair_fitter_mp on a NIfTI-1 image.
+    
+    Parameters
+    ----------
+    nii_in_fname : str
+        Input file path.
+        
+    nii_out_fname : str
```
Comment (salma1601) [inline sammba/modality_processors/perfusion_fair.py]: may be can be optional with default to None and then setting it to the input file path with a chosen suffix if not specified by the user

File: sammba/modality_processors/perfusion_fair.py [hunk 18]
```diff
@@ -0,0 +1,369 @@
...
+
+    nii_in = nib.load(nii_in_fname)
+    in_mat = nii_in.get_data()
+    all_s0 = in_mat.reshape((np.product(in_mat.shape[:-1]), in_mat.shape[-1]))
+    r = perf_fair_fitter_mp(all_s0, t1_blood, lambda_blood, ti, multiplier,
+                            t1_guess, picker_sel, picker_nonsel, ncpu)
+    r = np.array(r)
+    img = nib.Nifti1Image(np.reshape(r, in_mat.shape[:-1] + (r.shape[1],)),
+                          nii_in.get_affine())
+    img.to_filename(nii_out_fname)
```
Comment (salma1601) [inline sammba/modality_processors/perfusion_fair.py]: may be the output path can be returned

File: sammba/registration/struct.py [hunk 19]
```diff
@@ -320,8 +320,13 @@ def anats_to_common(anat_filenames, write_dir, brain_volume,
     out_undump = undump(in_file=out_tstat.outputs.out_file,
                         out_file=os.path.join(write_dir, 'undump.nii.gz'),
                         outputtype='NIFTI_GZ')
-    out_refit = refit(in_file=out_undump.outputs.out_file,
-                      xorigin='cen', yorigin='cen', zorigin='cen')
+    import subprocess
```
Comment (salma1601) [inline sammba/registration/struct.py]: I fixed this issue in PR #89. Please do not worry about this bug as this code will be rebased at some point on the fix

File: sammba/modality_processors/perfusion_fair.py (hunk 15 above)
Comment (nadkarni-na) [reply sammba/modality_processors/perfusion_fair.py]: agreed for a default value, though of mp.cpu_count() - 1, which seems to be typical practice

File: sammba/modality_processors/perfusion_fair.py (hunk 1 above)
Comment (nadkarni-na) [reply sammba/modality_processors/perfusion_fair.py]: OK

File: sammba/modality_processors/perfusion_fair.py (hunk 2 above)
Comment (nadkarni-na) [reply sammba/modality_processors/perfusion_fair.py]: It reduces the number of parameters entered further up the chain. In practice, the only function I want the user to touch is perf_fair_niiptbl_proc or maybe perf_fair_nii_proc. I would not be against giving all the other functions leading underscores

File: sammba/modality_processors/perfusion_fair.py (hunk 7 above)
Comment (nadkarni-na) [reply sammba/modality_processors/perfusion_fair.py]: It's a literature convention, the best that can be done is to provide default values

File: sammba/modality_processors/perfusion_fair.py (hunk 12 above)
Comment (nadkarni-na) [reply sammba/modality_processors/perfusion_fair.py]: It was quicker to write than changing the block indentation due to commenting out the previous line. Will delete it

File: sammba/modality_processors/perfusion_fair.py (hunk 13 above)
Comment (nadkarni-na) [reply sammba/modality_processors/perfusion_fair.py]: As before, it's a literature convention to use that particular output unit (same issue for input units); will provide default values

File: sammba/modality_processors/perfusion_fair.py (hunk 14 above)
Comment (nadkarni-na) [reply sammba/modality_processors/perfusion_fair.py]: Yep

File: sammba/modality_processors/perfusion_fair.py (hunk 17 above)
Comment (nadkarni-na) [reply sammba/modality_processors/perfusion_fair.py]: OK

File: sammba/modality_processors/perfusion_fair.py (hunk 18 above)
Comment (nadkarni-na) [reply sammba/modality_processors/perfusion_fair.py]: OK

**Comment by salma1601** [comment]
general questions:

- how do you handle spatial alignment and in particular the tagging impact on standard realignment procedure ?
- for which type of ASL labeling your computations apply ? is it FAIR PASL? and only this one ? can you give the main reference for it ?
- Can you give a paper reference for your quantification formulae ?

**Comment by nadkarni-na** [comment]
> how do you handle spatial alignment and in particular the tagging impact on standard realignment procedure ?

see also https://github.com/sammba-mri/sammba-mri/pull/84#issuecomment-364943295 and that conversation in general. quick version is the acquisition is too slow and sparse for spatial alignment to be used. even if it was, some of the tags are so attenuating that is would not work anyway (same for CEST). the acquisition assumes the animal does not move

> for which type of ASL labeling your computations apply ? is it FAIR PASL? and only this one ? can you give the main reference for it ?

only FAIR PASL, and only guaranteed for the Bruker implementation as we have no other manufacturer's data. Bruker is likely the only relevant implementation anyway as they are the only animal manufacturer, and the kind of FAIR we use is probably not done in humans due to high RF power. the best reference is actually the Paravision manual which references the papers they based the sequence on. I'll DM you the link to it on our server as the PDF is likely copyrighted/proprietary material that should not be shared in public

> see above answer, it's in the manual

File: sammba/modality_processors/perfusion_fair.py (hunk 9 above)
Comment (nadkarni-na) [reply sammba/modality_processors/perfusion_fair.py]: nope they're pandas dataseries

File: sammba/modality_processors/perfusion_fair.py (hunk 8 above)
Comment (nadkarni-na) [reply sammba/modality_processors/perfusion_fair.py]: OK

File: sammba/modality_processors/perfusion_fair.py (hunk 10 above)
Comment (nadkarni-na) [reply sammba/modality_processors/perfusion_fair.py]: it's useful for diagnostic use; in any case this function may be worth unexposing

File: sammba/modality_processors/perfusion_fair.py (hunk 11 above)
Comment (nadkarni-na) [reply sammba/modality_processors/perfusion_fair.py]: OK

**Comment by nadkarni-na** [comment]
Manual is at /home/Promane/2014-ROMANE/5_Experimental-Plan-Experiments-Results/mouse/equipment/Paravision6.0/english/pvman/ParaVision_6.0/ParaVision_6.0_en_536169227.pdf

Can't remember the page or section; if you search for "Kober" you should find it

File: sammba/modality_processors/perfusion_fair.py [hunk 20]
```diff
@@ -0,0 +1,387 @@
...
+        
+    t1_blood : int or float
+        T1 of blood in ms at the acquisition field strength. Empirical
+        constant; an example value is 2800 at 11.7T
+        (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3752414/table/T1,
+        Table 1 of Blood longitudinal (T1) and transverse (T2) relaxation time
+        constants at 11.7 Tesla, MAGMA. 2012 Jun; 25(3): 245–249,
+        Ai-Ling Lin, Qin Qin, Xia Zhao, and Timothy Q. Duong).
+
+    kwargs : dict, optional
```
Comment (salma1601) [inline sammba/modality_processors/perfusion_fair.py]: so if the user wants to know what these additional arguments are, he is directed to `perf_fair_nii_proc`, where some extra parameters are extracted but for others redirected to `perf_fair_fitter_mp`, and from there redirected to `perf_fair_fitter`. could you please try improve this ?

File: sammba/modality_processors/perfusion_fair.py [hunk 21]
```diff
@@ -0,0 +1,387 @@
...
+    r = np.array(r)
+    img = nib.Nifti1Image(np.reshape(r, in_mat.shape[:-1] + (r.shape[1],)),
+                          nii_in.get_affine())
+    
+    if nii_out_fname is None:
+        nii_out_fname = nii_in_fname.replace('.nii.gz', '_proc.nii.gz')
+    return img.to_filename(nii_out_fname)
+
+
+def perf_fair_niiptbl_proc(nii_in_fname, t1_blood, **kwargs):
```
Comment (salma1601) [inline sammba/modality_processors/perfusion_fair.py]: if this is the function you will expose to the user, I think you need to explain what is the output in its documentation. The extensive explanations (CBF etc) should go rather in this exposed function than in the others

**Comment by salma1601** [comment]
I tried on some data and it looks working. I like your progress bar, but I think we shall maintain the minimal dependency strategy

File: sammba/modality_processors/perfusion_fair.py [hunk 22]
```diff
@@ -0,0 +1,387 @@
...
+    # but with _ptbl.txt at the end rather than just .nii.gz. arguably this
+    # function should directly take the text file name as input, but it is
+    # difficult to envisage a scenario where this would be needed. the
+    # *_ptbl.txt file has a specific format created by
+    # sammba.io_conversions.dcm_to_nii. if someone supplies their own file
+    # name, it would mean they changed the *_ptbl.txt name (why?) or they have
+    # their own formatted text file which this function would not be able to
+    # process
+    
+    bfx = nii_fname.split('_')[-1].split('.')[0]
```
[the remark this answers is not in the document]
Comment (nadkarni-na) [reply sammba/modality_processors/perfusion_fair.py]: will add a comment

File: sammba/modality_processors/perfusion_fair.py [hunk 23]
```diff
@@ -0,0 +1,387 @@
...
+    long_ti:
+        full list of TIs, the same length and in the same order as the
+        acquisitions (the final dimension of the NIfTI-1 file)
+    fc:
+        full list of inversion type descriptions ("Selective" or
+        "Non-selective"), the same length and in the same order as the
+        acquisitions (the final dimension of the NIfTI-1 file). fc is a DICOM
+        acronym for frame comment, the DICOM field used for storing the
+        inversion type descriptions
+    picker_sel:
```
[the remark this answers is not in the document]
Comment (nadkarni-na) [reply sammba/modality_processors/perfusion_fair.py]: strictly speaking it's not the positions, it's a marker of them, hence picker

File: sammba/modality_processors/perfusion_fair.py [hunk 24]
```diff
@@ -0,0 +1,387 @@
...
+    ti:
+        simple list of all TIs
+    long_ti:
+        full list of TIs, the same length and in the same order as the
+        acquisitions (the final dimension of the NIfTI-1 file)
+    fc:
+        full list of inversion type descriptions ("Selective" or
+        "Non-selective"), the same length and in the same order as the
+        acquisitions (the final dimension of the NIfTI-1 file). fc is a DICOM
+        acronym for frame comment, the DICOM field used for storing the
```
[the remark this answers is not in the document]
Comment (nadkarni-na) [reply sammba/modality_processors/perfusion_fair.py]: expanding out acronyms makes code difficult to read.

File: sammba/modality_processors/perfusion_fair.py [hunk 25]
```diff
@@ -0,0 +1,387 @@
...
+    # difficult to envisage a scenario where this would be needed. the
+    # *_ptbl.txt file has a specific format created by
+    # sammba.io_conversions.dcm_to_nii. if someone supplies their own file
+    # name, it would mean they changed the *_ptbl.txt name (why?) or they have
+    # their own formatted text file which this function would not be able to
+    # process
+    
+    bfx = nii_fname.split('_')[-1].split('.')[0]
+    ptbl = pd.read_table(nii_fname.replace(bfx + '.nii.gz', bfx + '_ptbl.txt'))
+    ti = ptbl.TI[ptbl.slice == 1][ptbl.FC == 'Selective Inversion'].tolist()
```
[the remark this answers is not in the document]
Comment (nadkarni-na) [reply sammba/modality_processors/perfusion_fair.py]: they're not numpy arrays, they're pandas dataseries that have no inbuilt method to convert directly to a numpy array

File: sammba/modality_processors/perfusion_fair.py [hunk 26]
```diff
@@ -0,0 +1,387 @@
...
+    picker_sel = [n for n,x in enumerate(fc) if x == 'Selective Inversion']
+    picker_nonsel = [n for n,x in enumerate(fc) if x == 'Non-selective Inversion']
+    return {'TI':ti, 'long_TI':long_ti, 'FC':fc,
+            'picker_sel':picker_sel, 'picker_nonsel':picker_nonsel}
+
+
+# the perfusion fluid-attenuated inversion-recovery function
+# the jacobian should be callable but I do not know enough maths to create it
+# pars = parameters
+def _fair_t1_func(pars, s0, ti):
```
[the remark this answers is not in the document]
Comment (nadkarni-na) [reply sammba/modality_processors/perfusion_fair.py]: it's a coding convention used elsewhere: in fact sometimes they just use a fun_ prefix e.g. fun_fair_t1, but fun_fair sounds like the beginning of joke with no punchline, so a suffix is better

File: sammba/modality_processors/perfusion_fair.py [hunk 27]
```diff
@@ -0,0 +1,387 @@
...
+        
+    using scipy.optimize.least_squares (default Trust Region Reflective
+    algorithm). Bias, M0 and T1 are all estimated with a zero lower bound.
+    Starting values for bias (zero) and M0 (mean of input signals) are
+    calculated automatically; T1 has to be supplied, with the mean of TIs
+    being a good guess.
+    
+    Parameters
+    ----------
+    s0 : numpy array of int or float
```
[the remark this answers is not in the document]
Comment (nadkarni-na) [reply sammba/modality_processors/perfusion_fair.py]: it can handle both, so I say so

File: sammba/modality_processors/perfusion_fair.py [hunk 28]
```diff
@@ -0,0 +1,387 @@
...
+    (method=least_squares), whose results bizarrely look closer to those of R
+    minpack.lm nlsLM (to confirm), and are of course identical to
+    scipy.optimize.least_squares with method='trf' (default).
+    """
+    
+    return ls(_fair_t1_func, np.array([0, np.mean(s0), t1_guess]),
+              bounds=([0, 0, 0], np.inf), args=(s0, np.array(ti)))
+
+
+def perf_fair_fitter(s0, t1_blood, ti, t1_guess, picker_sel, picker_nonsel,
```
[the remark this answers is not in the document]
Comment (nadkarni-na) [reply sammba/modality_processors/perfusion_fair.py]: all instances of fitter are now fit

File: sammba/version.py [hunk 29]
```diff
@@ -53,6 +53,10 @@
         'min_version': '0.15.0',
         'required_at_installation': True,
         'install_info': _SAMMBA_INSTALL_MSG}),
+	('tqdm', {
```
[the remark this answers is not in the document]
Comment (nadkarni-na) [reply sammba/version.py]: It's good (often mandatory) in computer science and general design practice in cases where responses will take more than a few seconds, arguably less. e.g. https://developer.apple.com/macos/human-interface-guidelines/indicators/progress-indicators/ and https://shiny.rstudio.com/articles/progress.html

If there is any part of our code where there will plausibly be no CLI response for 3s, a progress indication should be added where possible if it does not already exist. There are several, e.g. DICOM to NIfTI conversion. That said, using this example, there are limits to what can be done; if the limiting step is handled by an outside process (in this case, dcmdump), not much can be done.

File: sammba/modality_processors/perfusion_fair.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (salma1601) [reply sammba/modality_processors/perfusion_fair.py]: I don't think so. I don't want to spend time guessing what the coder wants to output at each line. Moreover I personally find it hard to follow a code when I need to remember several acronyms because of the memory load.

File: sammba/modality_processors/perfusion_fair.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (salma1601) [reply sammba/modality_processors/perfusion_fair.py]: explicit naming is better than a comment: you don't need to search for the comment

File: sammba/modality_processors/perfusion_fair.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (salma1601) [reply sammba/modality_processors/perfusion_fair.py]: I think no pb if if it's just a marker.

File: sammba/modality_processors/perfusion_fair.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (salma1601) [reply sammba/modality_processors/perfusion_fair.py]: OK

File: sammba/modality_processors/perfusion_fair.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (salma1601) [reply sammba/modality_processors/perfusion_fair.py]: >coding convention used elsewhere

where ?
you mean you suffix with func because it is a function ?

File: sammba/modality_processors/perfusion_fair.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (salma1601) [reply sammba/modality_processors/perfusion_fair.py]: it is somehow expected that functions that work with floats still work when they are int. I think people don't explicit this

File: sammba/modality_processors/perfusion_fair.py (hunk 28 above)
[the remark this answers is not in the document]
Comment (salma1601) [reply sammba/modality_processors/perfusion_fair.py]: `fit` is somehow vague for me. Can't you find another name that gives more information about what the function does, e.g. `compute_cbf`.
Also, I think you don't really need to prefix with `perf_fair`, since your module is named perfusion_fair

File: sammba/modality_processors/perfusion_fair.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (nadkarni-na) [reply sammba/modality_processors/perfusion_fair.py]: it's on the same line, more efficient than explicit naming which would expand the variable name and clutter the code

**Comment by nadkarni-na** [comment]
any ideas on the jacobian?

File: sammba/version.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (salma1601) [reply sammba/version.py]: I do like printing out something if execution last long, or even using a progress bar. What I don't like is adding a dependency for that. We can have an utils function that prints a basic progress bar without any external dependency

**Comment by salma1601** [comment]
which jacobian?

File: sammba/version.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (nadkarni-na) [reply sammba/version.py]: tqdm is very mature, stable and efficient, it's not a big dependency

**Comment by nadkarni-na** [comment]
https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.least_squares.html see jac : {‘2-point’, ‘3-point’, ‘cs’, callable}, optional

File: sammba/modality_processors/perfusion_fair.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (nadkarni-na) [reply sammba/modality_processors/perfusion_fair.py]: yes, see scipy.optimize.least_squares examples

**Comment by salma1601** [comment]
@nadkarni-na I want to merge this PR but there is a rebase problem and tests are not passing becase of tqdm not being installed. Do you want to deal with this or do you prefer me fetching your branch and doing the changes ?

**Comment by nadkarni-na** [comment]
To confirm, that means adding python-tqdm [here](https://github.com/sammba-mri/sammba-mri/blob/339c20e2d4685465777894077ad8acfa8a70c793/continuous_integration/install.sh#L116)?

**Comment by salma1601** [comment]
yes and also for conda
https://github.com/sammba-mri/sammba-mri/blob/339c20e2d4685465777894077ad8acfa8a70c793/continuous_integration/install.sh#L42

**Comment by nadkarni-na** [comment]
still failss

**Comment by salma1601** [comment]
try add a version for tqdm in https://github.com/sammba-mri/sammba-mri/blob/master/.travis.yml#L29
as well as similar lines in the same file.
BTW I think of removing tqdm dependency at some point and may be using a more basic progress bar that does not need to add a dependency. But let's merge this first

**Comment by nadkarni-na** [comment]
It almost worked. The 356.4 failure I can solve easily. Not sure how to fix 356.2

**Comment by salma1601** [comment]
scipy.optimize.least_squares has been added in scipy 0.17.0, so it will not work for older versions. Isn't there a possible alternative ?

**Comment by nadkarni-na** [comment]
[lmfit](https://lmfit.github.io/lmfit-py/) which I have backup code for, but it is an unnecessary dependency given the later versions of scipy are capable of bounded non-linear optimisation. what's blocking upgrading scipy?

**Comment by salma1601** [comment]
what about scipy.optimize.leastsq ? 
Ubuntu Trusty 14.04 comes with scipy 0.13.3 and we need to support at least from 0.14

**Comment by nadkarni-na** [comment]
scipy.optimize.leastsq is not bounded

**Comment by salma1601** [comment]
what about scipy.optimize.fmin_slsqp ?
Otherwise I prefer you add lmfit. I will remove tqdm dependency at some point but optimization looks like the core work of this submodule so I am not against adding this dependency.

**Comment by nadkarni-na** [comment]
scipy.optimize.fmin_slsqp looked mathematically dubious so used lmfit.

the installation config files were already messy and are even more so now but they work.

**Comment by salma1601** [comment]
Ok merging. Feel free to improve the installation config files in a future PR.
